File: SecureBlackbox Samples/HTTP Server/c++/httpserver.cpp
/*
 * SecureBlackbox 2024 C++ Edition - Sample Project
 *
 * This sample project demonstrates the usage of SecureBlackbox in a 
 * simple, straightforward way. It is not intended to be a complete 
 * application. Error handling and other checks are simplified for clarity.
 *
 * www.nsoftware.com/secureblackbox
 *
 * This code is subject to the terms and conditions specified in the 
 * corresponding product license agreement which outlines the authorized 
 * usage and restrictions.
 */

#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include "../../include/secureblackbox.h"

class MyHTTPServer : public HTTPServer
{
public:

    int FireGetRequest(HTTPServerGetRequestEventParams* e) override {
        printf("Request received (page: %s)\r\n", e->URI);

        this->SetResponseStatus(e->ConnectionID, 200);
        this->SetResponseString(e->ConnectionID,
                                "<html><head><title>SecureBlackbox HTTP server</title></head><body><h1>Hello there!</h1></body></html>",
                                "text/html", "");

        e->Handled = true;

        return 0;
    }
};

namespace ArgParser {
	static char* optval(int argc, char** argv, const char* option) {
		for (int x = 0; x < argc - 1; x++) {
			if (!strcmp(argv[x], option)) {
				return argv[x + 1];
			}
		}
		return (char*)"";
	}
}

static bool optext(int argc, char** argv, const char* option) {
	for (int x = 0; x < argc; x++) {
		if (!strcmp(argv[x], option)) {
			return true;
		}
	}
	return false;
}

using namespace ArgParser;

void displayHelp() {
	printf(
		"NAME\n"
		"  httpserver -- SecureBlackbox HTTPServer Demo Application\n\n"
		"SYNOPSIS\n"
		"  httpserver <listening_port> [-cert certificate_file] [-certpass certificate_password]\n\n"
		"DESCRIPTION\n"
		"  HTTPServer demonstrates the usage of HTTPServer from SecureBlackbox.\n"
		"  The options are as follows:\n\n"
		"  -cert           The certificate used in server.\n\n"
		"  -certpass       The password for the certificate.\n\n"
		"EXAMPLES\n"
		"  httpserver 80 \n\n"
		"  httpserver 8080 -cert C:\\certs\\mycert.pfx -certpass mypassword\n\n"
	);
}

int main(int argc, char* argv[]) {
    MyHTTPServer httpserver;
	CertificateManager cm;

	// Validate input
	if (argc < 2) {
		displayHelp();
		goto done;
	}

    const int port = atoi(argv[1]);
    httpserver.SetPort(port);

	char* cert = optval(argc, argv, "-cert");
	char* certpass = optval(argc, argv, "-certpass");

	if (strcmp(cert, "")) {
		cm.ImportFromFile(cert, certpass);
		httpserver.SetTLSServerCertCount(1);
		httpserver.SetTLSServerCertHandle(0, cm.GetCertHandle());
	}

    if (httpserver.Start() == 0) {
        printf("HTTP server started on port %d. Press enter to stop server.\r\n", port);
        getchar();

        httpserver.Stop();

        printf("Server stopped. \n\n");
    } else {
        printf("Failed to run the server\n");
		printf("Error: [%i] %s\n\n", httpserver.GetLastErrorCode(), httpserver.GetLastError());
		goto done;
    }

done:
	printf("Press Enter to exit the demo.\n");
	getchar();
}

File: SecureBlackbox Samples/HTTP Server/net/httpserver.cs
/*
 * SecureBlackbox 2024 .NET Edition - Sample Project
 *
 * This sample project demonstrates the usage of SecureBlackbox in a 
 * simple, straightforward way. It is not intended to be a complete 
 * application. Error handling and other checks are simplified for clarity.
 *
 * www.nsoftware.com/secureblackbox
 *
 * This code is subject to the terms and conditions specified in the 
 * corresponding product license agreement which outlines the authorized 
 * usage and restrictions.
 * 
 */

﻿using System;
using _CA;
using nsoftware.SecureBlackbox;
using SBMath;

class httpserver
{
    private static string optval(string[] args, string option)
    {
        for (int x = 0; x < args.Length - 1; x++)
        {
            if (args[x].Equals(option, StringComparison.CurrentCultureIgnoreCase))
            {
                return args[x + 1];
            }
        }
        return "";
    }

    private static bool optext(string[] args, string option)
    {
        for (int x = 0; x < args.Length; x++)
        {
            if (args[x].Equals(option, StringComparison.CurrentCultureIgnoreCase))
            {
                return true;
            }
        }
        return false;
    }

    private static void displayHelp(string errMes)
    {
        Console.WriteLine(
            "NAME\n" +
            "  httpserver -- SecureBlackbox HTTPServer Demo Application\n\n" +
            "SYNOPSIS\n" +
            "  httpserver <listening_port> [-cert certificate_file] [-certpass certificate_password]\n\n" +
            "DESCRIPTION\n" +
            "  HTTPServer demonstrates the usage of HTTPServer from SecureBlackbox.\n" +
            "  The options are as follows:\n\n" +
            "  -cert           The certificate used in http server.\n\n" +
            "  -certpass       The password for the certificate.\n\n" +
            "EXAMPLES\n" +
            "  httpserver 80 \n\n" +
            "  httpserver 8080 -cert C:\\certs\\mycert.pfx -certpass mypassword\n"
        );

        if (errMes.Length > 0)
        {
            Console.WriteLine("Error: " + errMes);
            Console.WriteLine();
        }

        confirmExit();
    }

    private static void confirmExit()
    {
        Console.WriteLine("Press Enter to exit the demo.");
        Console.ReadLine();
    }

    private static HTTPServer server;

    private static void _server_OnGetRequest(object sender, HTTPServerGetRequestEventArgs e)
    {
        Console.WriteLine("Request received (page: " + e.URI + ")");

        try
        {
            server.SetResponseStatus(e.ConnectionID, 200);
            server.SetResponseString(e.ConnectionID,
                    "<html><head><title>SecureBlackbox HTTP server</title></head><body><h1>Hello there!</h1></body></html>",
                    "text/html", "");

            e.Handled = true;
        }
        catch (Exception ex)
        {
            Console.WriteLine("Error: " + ex.Message);
        }
    }

    static void Main(string[] args)
    {
        if (args.Length < 1)
        {
            displayHelp("listening_port is required.");
            return;
        }

        server = new HTTPServer();
        try
        {
            server.Port = int.Parse(args[0]);
        }
        catch (Exception ex)
        {
            displayHelp("Invalid port value");
            return;
        }

        try
        {
            server.OnGetRequest += _server_OnGetRequest;


            if (optext(args, "-cert"))
            {
                CertificateManager cm = new CertificateManager();
                cm.ImportFromFile(optval(args, "-cert"), optval(args, "-certpass"));
                server.TLSServerChain.Add(cm.Certificate);
            }

            server.Start();

            Console.WriteLine("HTTP server started on port " + server.Port + ". Press enter to stop server.");
            Console.ReadLine();

            server.Stop();

            Console.WriteLine("Server stopped.\n");

            confirmExit();
        }
        catch (Exception ex)
        {
            Console.WriteLine("Error: " + ex.Message);
        }
    }
}





class ConsoleDemo
{
  /// <summary>
  /// Takes a list of switch arguments or name-value arguments and turns it into a dictionary.
  /// </summary>
  public static System.Collections.Generic.Dictionary<string, string> ParseArgs(string[] args)
  {
    System.Collections.Generic.Dictionary<string, string> dict = new System.Collections.Generic.Dictionary<string, string>();

    for (int i = 0; i < args.Length; i++)
    {
      // Add a key to the dictionary for each argument.
      if (args[i].StartsWith("/"))
      {
        // If the next argument does NOT start with a "/", then it is a value.
        if (i + 1 < args.Length && !args[i + 1].StartsWith("/"))
        {
          // Save the value and skip the next entry in the list of arguments.
          dict.Add(args[i].ToLower().TrimStart('/'), args[i + 1]);
          i++;
        }
        else
        {
          // If the next argument starts with a "/", then we assume the current one is a switch.
          dict.Add(args[i].ToLower().TrimStart('/'), "");
        }
      }
      else
      {
        // If the argument does not start with a "/", store the argument based on the index.
        dict.Add(i.ToString(), args[i].ToLower());
      }
    }
    return dict;
  }
  /// <summary>
  /// Asks for user input interactively and returns the string response.
  /// </summary>
  public static string Prompt(string prompt, string defaultVal)
  {
    Console.Write(prompt + (defaultVal.Length > 0 ? " [" + defaultVal + "]": "") + ": ");
    string val = Console.ReadLine();
    if (val.Length == 0) val = defaultVal;
    return val;
  }
}